File: Do_Not_Tap/core/src/com/nopalsoft/donttap/screens/Screens.java
package com.nopalsoft.donttap.screens;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputAdapter;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.nopalsoft.donttap.Assets;
import com.nopalsoft.donttap.DoNotTapGame;
import com.nopalsoft.donttap.Settings;
import com.nopalsoft.donttap.game.GameScreen;

public abstract class Screens extends InputAdapter implements Screen {
    public static final int SCREEN_WIDTH = 480;
    public static final int SCREEN_HEIGHT = 800;

    public static final float WORLD_WIDTH = 480;
    public static final float WORLD_HEIGHT = 800 - 80;// Minus 80 is the bookmarks bar

    public DoNotTapGame game;

    public OrthographicCamera camera;
    public SpriteBatch batch;
    public Stage stage;

    public Screens(final DoNotTapGame game) {
        this.stage = game.stage;
        this.stage.clear();
        this.batch = game.batch;
        this.game = game;

        camera = new OrthographicCamera(SCREEN_WIDTH, SCREEN_HEIGHT);
        camera.position.set(SCREEN_WIDTH / 2f, SCREEN_HEIGHT / 2f, 0);

        InputMultiplexer input = new InputMultiplexer(this, stage);
        Gdx.input.setInputProcessor(input);
    }

    @Override
    public void render(float delta) {
        if (delta > .1f)
            delta = .1f;

        update(delta);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        camera.update();
        batch.setProjectionMatrix(camera.combined);
        draw(delta);

        stage.act(delta);
        stage.draw();
    }

    Image blackFadeOut;

    public void changeScreenWithFadeOut(final Class<?> newScreen,
                                        final DoNotTapGame game) {
        changeScreenWithFadeOut(newScreen, game, -1);
    }

    public void changeScreenWithFadeOut(final Class<?> newScreen,
                                        final DoNotTapGame game, final int mode) {
        blackFadeOut = new Image(Assets.pixelNegro);
        blackFadeOut.setSize(SCREEN_WIDTH, SCREEN_HEIGHT);
        blackFadeOut.getColor().a = 0;
        blackFadeOut.addAction(Actions.sequence(Actions.fadeIn(.5f),
                Actions.run(new Runnable() {
                    @Override
                    public void run() {
                        if (newScreen == SelectScreenScreen.class)
                            game.setScreen(new SelectScreenScreen(game));
                        else if (newScreen == GameScreen.class)
                            game.setScreen(new GameScreen(game, mode));
                        else if (newScreen == MainMenuScreen.class)
                            game.setScreen(new MainMenuScreen(game));
                    }
                })));
        stage.addActor(blackFadeOut);
    }

    public void addPressEffect(final Actor actor) {
        actor.addListener(new InputListener() {
            @Override
            public boolean touchDown(InputEvent event, float x, float y,
                                     int pointer, int button) {
                actor.setPosition(actor.getX(), actor.getY() - 3);
                event.stop();
                return true;
            }

            @Override
            public void touchUp(InputEvent event, float x, float y,
                                int pointer, int button) {
                actor.setPosition(actor.getX(), actor.getY() + 3);
            }
        });
    }

    protected void addBackGround() {
        Table tabBackground = new Table();
        tabBackground.setFillParent(true);
        tabBackground.defaults().expand().uniform().fill();

        for (int col = 0; col < 4; col++) {

            int tileColorPos = MathUtils.random(3);
            for (int ren = 0; ren < 4; ren++) {
                Image img = new Image(Assets.tileBlanco);

                if (tileColorPos == ren) {

                    int colorTile = MathUtils.random(4);
                    switch (colorTile) {
                        case 0:
                            img = new Image(Assets.tileAmarillo);
                            break;
                        case 1:
                            img = new Image(Assets.tileAzul);
                            break;
                        case 2:
                            img = new Image(Assets.tileRojo);
                            break;
                        case 3:
                            img = new Image(Assets.tileMorado);
                            break;
                        case 4:
                            img = new Image(Assets.tileNaranja);
                            break;
                    }
                }
                tabBackground.add(img);
            }
            tabBackground.row();
        }
        stage.addActor(tabBackground);
    }

    public abstract void draw(float delta);

    public abstract void update(float delta);

    @Override
    public void resize(int width, int height) {
        stage.getViewport().update(width, height, true);
    }

    @Override
    public void show() {
    }

    @Override
    public void hide() {
        Settings.save();
    }

    @Override
    public void pause() {
    }

    @Override
    public void resume() {
    }

    @Override
    public void dispose() {
        stage.dispose();
        batch.dispose();
    }
}

File: Do_Not_Tap/core/src/com/nopalsoft/donttap/screens/Screens.kt
package com.nopalsoft.donttap.screens

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.InputAdapter
import com.badlogic.gdx.InputMultiplexer
import com.badlogic.gdx.Screen
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.InputListener
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.nopalsoft.donttap.Assets
import com.nopalsoft.donttap.DoNotTapGame
import com.nopalsoft.donttap.Settings.save
import com.nopalsoft.donttap.game.GameScreen

abstract class Screens(game: DoNotTapGame) : InputAdapter(), Screen {
    @JvmField
    var game: DoNotTapGame?

    var camera: OrthographicCamera
    var batch: SpriteBatch

    @JvmField
    var stage: Stage

    override fun render(delta: Float) {
        var delta = delta
        if (delta > .1f) delta = .1f

        update(delta)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)

        camera.update()
        batch.setProjectionMatrix(camera.combined)
        draw(delta)

        stage.act(delta)
        stage.draw()
    }

    var blackFadeOut: Image? = null

    init {
        this.stage = game.stage
        this.stage.clear()
        this.batch = game.batch
        this.game = game

        camera = OrthographicCamera(SCREEN_WIDTH.toFloat(), SCREEN_HEIGHT.toFloat())
        camera.position.set(SCREEN_WIDTH / 2f, SCREEN_HEIGHT / 2f, 0f)

        val input = InputMultiplexer(this, stage)
        Gdx.input.inputProcessor = input
    }

    @JvmOverloads
    fun changeScreenWithFadeOut(
        newScreen: Class<*>?,
        game: DoNotTapGame, mode: Int = -1
    ) {
        blackFadeOut = Image(Assets.pixelNegro)
        blackFadeOut!!.setSize(SCREEN_WIDTH.toFloat(), SCREEN_HEIGHT.toFloat())
        blackFadeOut!!.getColor().a = 0f
        blackFadeOut!!.addAction(
            Actions.sequence(
                Actions.fadeIn(.5f),
                Actions.run(object : Runnable {
                    override fun run() {
                        if (newScreen == SelectScreen::class.java) game.setScreen(SelectScreen(game))
                        else if (newScreen == GameScreen::class.java) game.setScreen(GameScreen(game, mode))
                        else if (newScreen == MainMenuScreen::class.java) game.setScreen(MainMenuScreen(game))
                    }
                })
            )
        )
        stage.addActor(blackFadeOut)
    }

    fun addPressEffect(actor: Actor) {
        actor.addListener(object : InputListener() {
            override fun touchDown(
                event: InputEvent, x: Float, y: Float,
                pointer: Int, button: Int
            ): Boolean {
                actor.setPosition(actor.getX(), actor.getY() - 3)
                event.stop()
                return true
            }

            override fun touchUp(
                event: InputEvent?, x: Float, y: Float,
                pointer: Int, button: Int
            ) {
                actor.setPosition(actor.getX(), actor.getY() + 3)
            }
        })
    }

    protected fun addBackGround() {
        val tabBackground = Table()
        tabBackground.setFillParent(true)
        tabBackground.defaults().expand().uniform().fill()

        for (col in 0..3) {
            val tileColorPos = MathUtils.random(3)
            for (ren in 0..3) {
                var img = Image(Assets.tileBlanco)

                if (tileColorPos == ren) {
                    val colorTile = MathUtils.random(4)
                    when (colorTile) {
                        0 -> img = Image(Assets.tileAmarillo)
                        1 -> img = Image(Assets.tileAzul)
                        2 -> img = Image(Assets.tileRojo)
                        3 -> img = Image(Assets.tileMorado)
                        4 -> img = Image(Assets.tileNaranja)
                    }
                }
                tabBackground.add<Image?>(img)
            }
            tabBackground.row()
        }
        stage.addActor(tabBackground)
    }

    abstract fun draw(delta: Float)

    abstract fun update(delta: Float)

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun show() {
    }

    override fun hide() {
        save()
    }

    override fun pause() {
    }

    override fun resume() {
    }

    override fun dispose() {
        stage.dispose()
        batch.dispose()
    }

    companion object {
        const val SCREEN_WIDTH: Int = 480
        const val SCREEN_HEIGHT: Int = 800

        const val WORLD_WIDTH: Float = 480f
        val WORLD_HEIGHT: Float = (800 - 80 // Minus 80 is the bookmarks bar
                ).toFloat()
    }
}
